btree/btree.c

```diff
@@ -2952,6 +2961,539 @@ static void ff_node_free_credits(const struct nd *node,
+{
+	/**
+	 * @brief This function will fill the provided slot with its key and
+	 *        value based on the index provided.
+	 *        vkvv_key() functions should be used to get the
+	 *        memory address pointing to key and val.
+	 *        vkvv_keysize() functions should be used to
+	 *        fill the v_count parameter of the slot vector.
+	 */
+}
+
+static void vkvv_child(struct slot *slot, struct segaddr *addr)
+{
+	/**
+	 * @brief This function will return the memory address pointing to its
+	 *        child node. We will call the function vkvv_val() for this
+	 *        purpose.
+	 *        This function is called for internal nodes.
+	 */
+}
+
+static bool vkvv_isfit(struct slot *slot)
+{
+	/**
+	 * @brief This function will determine if the current key and value
+	 *        can be added to the node.
+	 *        It also checks the validity of the record.
+	 *        Modifications will be required in nt_isfit() function
+	 *        declaration as we need to have the sizes of incoming key and
+	 *        value.
+	 *
+	 *        keysize + valsize <= vkvv_space(slot->s_node)
+	 */
+}
+
+static void vkvv_done(struct slot *slot, struct m0_be_tx *tx, bool modified)
+{
+
+}
+
+static void vkvv_make(struct slot *slot, struct m0_be_tx *tx)
+{
+	/**
+	 * @brief This function will create space to add a new entry at the
+	 *        given index. It is assumed that the possibility whether this
+	 *        can occur or not is determined before calling this function.
+	 *
+	 *        We will maintain an extra record than vkvv_used in the
+	 *        directory. Maintaining this extra offset value will help in
+	 *        size calculations.
+	 *
+	 *        start_key_addr = vkvv_key(slot->s_node, index)
+	 *        start_val_addr = vkvv_val(slot->s_node, index)
+	 *        count = num of records i.e. vkvv_used
+	 *
+	 *        if(index==count) {
+	 *          //No need to do memmove() as data will added to the end of
+	 *          //current series of keys and values. Just update the
+	 *          //directory to keep a record of the next possible offset.
+	 *          dir_rec[index+1].key_offset = dir_rec[index].key_offset + incoming_key_size
+	 *          dir_rec[index+1].val_offset = dir_rec[index].val_offset + incoming_val_size
+	 *
+	 *        } else {
+	 *
+	 *          total_key_size = dir_rec[count+1].key_offset - dir_rec[index].key_offset
+	 *          total_val_size = dir_rec[count+1].val_offset - dir_rec[index].val_offset
+	 *
+	 *          //Update directory
+	 *          idx = index
+	 *          temp_count = count
+	 *          while(temp_count >= idx) {
+	 *            dir_rec[temp_count+1].key_offset = dir_rec[temp_count].key_offset
+	 *            dir_rec[temp_count+1].val_offset = dir_rec[temp_count].val_offset
+	 * 
+	 *            dir_rec[temp_count+1].key_offset = dir_rec[temp_count+1].key_offset + incoming_key_size
+	 *            dir_rec[temp_count+1].val_offset = dir_rec[temp_count+1].val_offset + incoming_value_size
```

Comment:
It should be dir_rec[index+1].val_offset = dir_rec[index].val_offset - incoming_val_size
Please check

Reply:
same comment as above.